In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib
matplotlib.rcParams.update({'errorbar.capsize': 2})
np.set_printoptions(suppress=True) #sopprime la notazione scientifica

In [2]:
n = 10 #numero di misure di tensione
seed = 2 #il seed
R_nominale = 200 #valore della resistenza

In [3]:
def mean(lista):
    media = 0
    n = len(lista)
    for i in range(n):
        media += lista[i]
    media = media/n
    return media

In [4]:
def st_dev(lista):
    n = len(lista)
    l = 0
    media = mean(lista)
    for i in range(n):
        l += (lista[i] - media)**2
    #print(l)
    l = l/n
    #print(l)
    std = np.sqrt(l)
    std_media = std/ np.sqrt(n) 
    return std, std_media

In [5]:
def mc_simulation(n,seed, R_nominale, start = 1.0, passo = 1.0):
    np.random.seed(seed)
    R = np.zeros(n)
    #print(R)
    ddp = np.arange(start,(n*passo)+start, passo) #vettore ddp 
    #print(ddp)
    I = ddp/R_nominale # V = IR -> I = V/R
    #print(I)

    for i in range(n):
        #voltmetro 3 % supponiamo incertezze gaussiana
        u1,u2 = np.random.rand(2)
        g1 = np.sqrt(-2*np.log(u1))*np.cos(2*np.pi*u2)
        g2 = np.sqrt(-2*np.log(u1))*np.sin(2*np.pi*u2)
        #print(g1,g2)
        ddp[i] += 0.05*g1
        #print(ddp[i])
        I[i] += I[i]*0.05*g2
        R[i] = ddp[i]/I[i]

    df = pd.DataFrame(data={'ddp (V)': ddp,'I (A)': I, 'R (Ohm)': R})
    print(df)
    return ddp, I, R

In [6]:
print("(a) tabulare le 10 coppie di tensione e corrente ottenute, e il valore di resistenza corrispondente a ogni coppia \n")
ddp, I, R = mc_simulation(n, seed, R_nominale)

(a) tabulare le 10 coppie di tensione e corrente ottenute, e il valore di resistenza corrispondente a ogni coppia 

     ddp (V)     I (A)     R (Ohm)
0   1.063572  0.005052  210.514921
1   1.949754  0.010216  190.848662
2   2.968167  0.015864  187.097698
3   3.934797  0.018787  209.447021
4   4.991808  0.026930  185.363827
5   5.952020  0.029733  200.179241
6   6.900224  0.034701  198.846012
7   8.020244  0.036413  220.258699
8   8.971925  0.045046  199.173525
9  10.025321  0.050692  197.767971


In [7]:
print("(b) calcolare la resistenza media, la sua deviazione standard e la deviazione standard della media e verificare la compatibilità della media con il valore nominale di resistenza. \n")

media = mean(R)
sd, sd_media= st_dev(R)
print(('La media di R è %10.7f Ohm, con sd %10.7f con sd dalla media %10.7f')
      %(media,sd, sd_media))
print(('Il valore di R è (%.1f +- %.1f) Ohm')%(media, sd_media))

(b) calcolare la resistenza media, la sua deviazione standard e la deviazione standard della media e verificare la compatibilità della media con il valore nominale di resistenza. 

La media di R è 199.9497577 Ohm, con sd 10.4002991 con sd dalla media  3.2888633
Il valore di R è (199.9 +- 3.3) Ohm


In [8]:
print("(c) discutere la propagazione degli errori e cosa cambia ripetendo la simulazione con 100 coppie di misure \n")

ddp_2, I_2, R_2 = mc_simulation(n = 100, seed = seed, R_nominale = R_nominale, start = 1.0, passo = 1.0)

media_2 = mean(R_2)
sd_2, sd_media_2 = st_dev(R_2)
print(('La media di R è %10.7f Ohm, con sd %10.7f con sd dalla media %10.7f')
      %(media_2,sd_2, sd_media_2))
print(('Il valore di R è (%.1f +- %.1f) Ohm')%(media_2, sd_media_2))

(c) discutere la propagazione degli errori e cosa cambia ripetendo la simulazione con 100 coppie di misure 

      ddp (V)     I (A)     R (Ohm)
0    1.063572  0.005052  210.514921
1    1.949754  0.010216  190.848662
2    2.968167  0.015864  187.097698
3    3.934797  0.018787  209.447021
4    4.991808  0.026930  185.363827
..        ...       ...         ...
95  96.014035  0.485034  197.953109
96  96.990578  0.506548  191.473626
97  97.931572  0.518334  188.935086
98  98.982262  0.490850  201.654896
99  99.993839  0.490040  204.052579

[100 rows x 3 columns]
La media di R è 198.9254684 Ohm, con sd  9.2932911 con sd dalla media  0.9293291
Il valore di R è (198.9 +- 0.9) Ohm


(a) tabulare le 10 coppie di tensione e corrente ottenute, e il valore di resistenza corrispondente a ogni coppia 

     ddp (V)     I (A)     R (Ohm)
0   1.063572  0.005052  210.514921
1   1.949754  0.010216  190.848662
2   2.968167  0.015864  187.097698
3   3.934797  0.018787  209.447021
4   4.991808  0.026930  185.363827
5   5.952020  0.029733  200.179241
6   6.900224  0.034701  198.846012
7   8.020244  0.036413  220.258699
8   8.971925  0.045046  199.173525
9  10.025321  0.050692  197.767971



(b) calcolare la resistenza media, la sua deviazione standard e la deviazione standard della media e verificare la compatibilità della media con il valore nominale di resistenza. 

La media di R è 199.9497577 Ohm, con sd 10.4002991 con sd dalla media  3.2888633
Il valore di R è (200 +- 3) Ohm

Riporto sopra il valore best della resistenza R con errore sigma associato opportunamente approssimato. Calcolando l'errore massimo associato, moltiplicando la sigma per 3, otteniamo che:

R = (200 +- 10) Ohm 

Possiamo affermare che la discrepanza non è significativa. Dato che appunto stiamo parlando di errore massimo, possiamo dire di avere il 100% che la discrepanza stessa cada entro questo intervallo di confidenza.



(c) discutere la propagazione degli errori e cosa cambia ripetendo la simulazione con 100 coppie di misure

      ddp (V)     I (A)     R (Ohm)
0    1.063572  0.005052  210.514921
1    1.949754  0.010216  190.848662
2    2.968167  0.015864  187.097698
3    3.934797  0.018787  209.447021
4    4.991808  0.026930  185.363827
..        ...       ...         ...
95  96.014035  0.485034  197.953109
96  96.990578  0.506548  191.473626
97  97.931572  0.518334  188.935086
98  98.982262  0.490850  201.654896
99  99.993839  0.490040  204.052579

La media di R è 198.9254684 Ohm, con sd  9.2932911 con sd dalla media  0.9293291
Il valore di R è (198.9 +- 0.9) Ohm

Osservando il valore best della resistenza R con errore sigma associato, come riportato sopra opportunamente approssimato. Calcolando l'errore massimo associato, moltiplicando la sigma per 3 e approssimando opportunamente, otteniamo che: 

R = (199 +- 3) Ohm 

Similmente al punto c) notiamo che la discrepanza non è significativa. Infatti, possiamo dire anche in questo caso di avere il 100% che il valore di R cada entro l'intervallo di confidenza.

In questo caso, possiamo notare che, come ci si può aspettare, all'aumentare del numero di misure, in questo caso specificamente 100, abbiamo una maggiore precisione sul valore di R.